cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/DBRecord.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.templates.etl.common;
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import com.google.common.collect.Lists;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.lib.db.DBWritable;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Writable class for DB Source/Sink
+ *
+ * {@see DBWritable}, {@see DBInputFormat}, {@see DBOutputFormat}
+ */
+public class DBRecord implements Writable, DBWritable {
+
+  private StructuredRecord record;
+
+  public void readFields(DataInput in) throws IOException {
+  }
+
+  public StructuredRecord getRecord() {
+    return record;
+  }
+
+  public void readFields(ResultSet resultSet) throws SQLException {
+    List<Schema.Field> fields = Lists.newArrayList();
+    ResultSetMetaData metadata = resultSet.getMetaData();
+    // ResultSetMetadata columns are numbered starting with 1
+    for (int i = 1; i <= metadata.getColumnCount(); i++) {
+      String columnName = metadata.getColumnName(i);
+      String columnType = metadata.getColumnClassName(i);
+      Schema.Field field = Schema.Field.of(columnName, Schema.of(getType(columnType)));
+      fields.add(field);
+    }
+    Schema schema = Schema.recordOf("dbRecord", fields);
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
+    for (Schema.Field field : fields) {
+      recordBuilder.set(field.getName(), resultSet.getObject(field.getName()));
+    }
+    record = recordBuilder.build();
+  }
+
+  private Schema.Type getType(String columnType) {
+    if (columnType == null) {
+      return Schema.Type.NULL;
+    } else if (String.class.getName().equals(columnType)) {
+      return Schema.Type.STRING;
+    } else if (Boolean.class.getName().equals(columnType)) {
+      return Schema.Type.BOOLEAN;
+    } else if (Integer.class.getName().equals(columnType)) {
+      return Schema.Type.INT;
+    } else if (Long.class.getName().equals(columnType)) {
+      return Schema.Type.LONG;
+    } else if (Float.class.getName().equals(columnType)) {
+      return Schema.Type.FLOAT;
+    } else if (Double.class.getName().equals(columnType)) {
+      return Schema.Type.DOUBLE;
+    } else {
+      throw new IllegalArgumentException("Unsupported datatype: " + columnType);
+    }
+  }
+
+  public void write(DataOutput out) throws IOException {
```

Comment:
So I presume that since we're writing to a DB, only implementing the ``write(PreparedStatement)`` method should be good enough. Other writables that directly extend from ``Writable`` have to implement this method, to write to the ``DataOutput``. I've tested that this works in unit test and standalone, and am about to test it on a cluster.